### PR TITLE
codeAction/resolve support for qute-ls

### LINKED ISSUE #604: Use resolve for code action

### PR DESCRIPTION
**Pull request by datho7561**
Closes #604

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/api/QuteTemplateCodeActionResolveProvider.java [hunk 1]
```diff
@@ -17,13 +17,22 @@
...
- * Request additional information from the associated java language server in
- * order to resolve the workspace edit to create a field, getter, or template
- * extension for a given type.
+ * Provides information from the associated java language server in order to
+ * resolve code actions.
  *
  * @author datho7561
  */
-public interface QuteTemplateGenerateMissingJavaMember {
+public interface QuteTemplateCodeActionResolveProvider {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/api/QuteTemplateCodeActionResolveProvider.java:25]: I find QuteTemplateCodeActionResolveProvider is difficult to understand that, it is a provider for textEdit to update Java file and not a code action

What about `QuteTemplateJavaTextEditProvider` ?

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/CodeActionResolverKind.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+package com.redhat.qute.services.codeactions;
+
+/**
+ * Represents what resolver is needed to resolve a code action
+ * 
+ * @author datho7561
+ */
+public enum CodeActionResolverKind {
+
+	GenerateMissingMember(1), //
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/CodeActionResolverKind.java:21]: `GenerateMissingMember(1);`

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/ICodeActionResolver.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.qute.services.codeactions;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.lsp4j.CodeAction;
+
+interface ICodeActionResolver {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/ICodeActionResolver.java:18]: Rename with `CodeActionResolver`

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/ICodeActionResolver.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package com.redhat.qute.services.codeactions;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.lsp4j.CodeAction;
+
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/ICodeActionResolver.java:17]: Add Javadoc

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionResolverForGenerateMissingMember.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.eclipse.lsp4j.CodeAction;
+
+import com.redhat.qute.commons.GenerateMissingJavaMemberParams;
+import com.redhat.qute.ls.api.QuteTemplateCodeActionResolveProvider;
+import com.redhat.qute.utils.JSONUtility;
+
+/**
+ * Resolves generate missing member code actions.
+ *
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionResolverForGenerateMissingMember.java:24]: Could add the JSON of a code action with data as comment

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionResolverForGenerateMissingMember.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+	private final QuteTemplateCodeActionResolveProvider resolverProvider;
+
+	public QuteCodeActionResolverForGenerateMissingMember(QuteTemplateCodeActionResolveProvider resolverProvider) {
+		this.resolverProvider = resolverProvider;
+	}
+
+	@Override
+	public CompletableFuture<CodeAction> resolveCodeAction(CodeAction codeAction) {
+		CodeActionUnresolvedData unresolvedData = JSONUtility.toModel(codeAction.getData(),
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionResolverForGenerateMissingMember.java:37]: Could you add as comment a JSON of codeAction.getData

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionResolverForGenerateMissingMember.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+				CodeActionUnresolvedData.class);
+		GenerateMissingJavaMemberParams params = JSONUtility.toModel(unresolvedData.getResolverData(),
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionResolverForGenerateMissingMember.java:39]: Could you add as comment a JSON of unresolvedData.getResolverData()

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteTemplateCodeActionResolvers.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		ICodeActionResolver resolver = resolvers.get(resolverKind);
+		if (resolver != null) {
+			return resolver.resolveCodeAction(unresolved);
+		}
+
+		LOGGER.severe(String.format("Cannot resolve CodeAction of kind %s", resolverKind));
+		return null;
+	}
+
+	private void initalizeResolvers(QuteTemplateCodeActionResolveProvider resolveProvider) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteTemplateCodeActionResolvers.java:66]: Use a synchronized method here and check at first if resolvers is != null

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/QuteTextDocumentService.java [hunk 9]
```diff
@@ -254,6 +257,17 @@ public CompletableFuture<List<InlayHint>> inlayHint(InlayHintParams params) {
 		}
 		return CompletableFuture.completedFuture(null);
 	}
+	
+	@Override
+	public CompletableFuture<CodeAction> resolveCodeAction(CodeAction codeAction) {
+		CodeActionUnresolvedData data = JSONUtility.toModel(codeAction.getData(), CodeActionUnresolvedData.class);
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/QuteTextDocumentService.java:263]: Could you add a JSON sample of data

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesConstants.java [hunk 10]
```diff
@@ -68,5 +71,20 @@ private ServerCapabilitiesConstants() {
...
 			Arrays.asList("{", "@", "#", ".", ":"));
 	public static final DocumentLinkOptions DEFAULT_DOCUMENT_LINK_OPTIONS = new DocumentLinkOptions(true);
+	public static final CodeActionOptions DEFAULT_CODE_ACTION_OPTIONS = new CodeActionOptions();
+	static {
+		List<String> allCodeActionKinds = Arrays.asList( //
+				CodeActionKind.QuickFix, //
+				CodeActionKind.Empty, //
+				CodeActionKind.Refactor, //
+				CodeActionKind.RefactorExtract, //
+				CodeActionKind.RefactorInline, //
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesConstants.java:81]: Are you sure that we support all CodeActionKind like refactor, etc 

Today we have just QuickFix

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesInitializer.java [hunk 11]
```diff
@@ -46,7 +47,8 @@ public static ServerCapabilities getNonDynamicServerCapabilities(ClientCapabilit
 		serverCapabilities
 				.setDocumentSymbolProvider(!clientCapabilities.isDocumentSymbolDynamicRegistrationSupported());
 		serverCapabilities.setHoverProvider(!clientCapabilities.isHoverDynamicRegistered());
-		serverCapabilities.setCodeActionProvider(!clientCapabilities.isCodeActionDynamicRegistered());
+		if (!clientCapabilities.isCodeActionDynamicRegistered())
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesInitializer.java:50]: Use bracket

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesInitializer.java [hunk 12]
```diff
@@ -46,7 +47,8 @@ public static ServerCapabilities getNonDynamicServerCapabilities(ClientCapabilit
 		serverCapabilities
 				.setDocumentSymbolProvider(!clientCapabilities.isDocumentSymbolDynamicRegistrationSupported());
 		serverCapabilities.setHoverProvider(!clientCapabilities.isHoverDynamicRegistered());
-		serverCapabilities.setCodeActionProvider(!clientCapabilities.isCodeActionDynamicRegistered());
+		if (!clientCapabilities.isCodeActionDynamicRegistered())
+			serverCapabilities.setCodeActionProvider(DEFAULT_CODE_ACTION_OPTIONS);
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesInitializer.java:51]: You need too to set DEFAULT_CODE_ACTION_OPTIONS in https://github.com/redhat-developer/quarkus-ls/blob/8bd86c7d1bb1aef8948803ee24754e3eb1dd7006/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/QuteCapabilityManager.java#L83

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesConstants.java (hunk 10 above)
Comment (datho7561) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/capabilities/ServerCapabilitiesConstants.java:81]: Okay. I'll remove everything except QuickFix and Empty

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java [hunk 13]
```diff
@@ -75,13 +75,13 @@ public QuteCodeActions(JavaDataModelCache javaCache) {
...
 	public CompletableFuture<List<CodeAction>> doCodeActions(Template template, CodeActionContext context, Range range,
-			QuteTemplateGenerateMissingJavaMember resolver, SharedSettings sharedSettings) {
+			QuteTemplateJavaTextEditProvider resolver, SharedSettings sharedSettings) {
 		List<CodeAction> codeActions = new ArrayList<>();
 		List<Diagnostic> diagnostics = context.getDiagnostics();
 		if (diagnostics == null || diagnostics.isEmpty()) {
 			return CompletableFuture.completedFuture(Collections.emptyList());
 		}
-
+
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java:84]: remove this extra spaces

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteLanguageService.java [hunk 14]
```diff
@@ -97,7 +100,7 @@ public QuteLanguageService(JavaDataModelCache javaCache) {
 	 * @return the CodeAction(s) to be added to the Qute template.
 	 */
 	public CompletableFuture<List<CodeAction>> doCodeActions(Template template, CodeActionContext context,
-			QuteTemplateGenerateMissingJavaMember resolver, Range range, SharedSettings sharedSettings) {
+			QuteTemplateJavaTextEditProvider resolver, Range range, SharedSettings sharedSettings) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteLanguageService.java:103]: Rename resolver with javaTextEditProvidzr

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/CodeActionRequest.java [hunk 15]
```diff
@@ -59,7 +59,7 @@ public Diagnostic getDiagnostic() {
 		return diagnostic;
 	}
 
-	public QuteTemplateGenerateMissingJavaMember getResolver() {
+	public QuteTemplateJavaTextEditProvider getResolver() {
 		return resolver;
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/CodeActionRequest.java:63]: Rename with getJavaTextEditProvider

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionForUnknownProperty.java [hunk 16]
```diff
@@ -68,7 +68,7 @@ public void doCodeActions(CodeActionRequest request, List<CompletableFuture<Void
 			}
 			Template template = request.getTemplate();
 			Diagnostic diagnostic = request.getDiagnostic();
-			QuteTemplateGenerateMissingJavaMember resolver = request.getResolver();
+			QuteTemplateJavaTextEditProvider resolver = request.getResolver();
 			SharedSettings sharedSettings = request.getSharedSettings();
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/codeactions/QuteCodeActionForUnknownProperty.java:72]: Rename resolver with javaTextEditProvider and do that for all classes

**Comment by angelozerr** [comment]
@datho7561 your contribution looks very great. The only thing that it should be nice is to rename resolver field, parameter, method with javaTextEditProvider for all classes which is linked with QuteTemplateJavaTextEditProvider

**Comment by angelozerr** [comment]
The resolve support works like a charm, great job @datho7561 !

**Pull request merged**